### ISSUE TITLE
Suspected memory leak when getting commit patch

### ISSUE THREAD

**Issue by MasonXon** [go-git/go-git#1117]
<img width="1236" alt="image" src="https://github.com/go-git/go-git/assets/86297429/cdaa3df1-4789-4796-b100-b2232b6cf605">
The phenomenon is as shown in the figure. I need to get the diff of each file in each commit in the git repo, but I found that a lot of memory is consumed when calculating the patch. Is there a problem with the processing logic here?

**Comment by onee-only** [comment]
Hi @MasonXon, I had a look at the source code from your memory profile. (thanks for sharing it 👍 )

Here is what I guess:

`filesystem.(*ObjectStorage)` caches packfile internally using `packfileFromCache` and `storePackfileInCache`. 

But if **MaxOpenDescriptors** option is not set (meaning 0) on ObjectStorage, it skips caching. 
https://github.com/go-git/go-git/blob/55f5d832fad510fecbad3fb413875d969e1441fb/storage/filesystem/object.go#L248-L250

So everytime it calls `(*ObjectStorage).packfile()`, it will open a packfile and create a new scanner.
Creating a new scanner is expensive because it calls `bufio.NewScanner` which allocates 4096 bytes.

And this should be the reason why it consumes a lot of memory while calculating the patch.

It might not be correct because I haven't test it yet.

**Comment by MasonXon** [comment]
I discovered this problem when traversing multiple large git repos. I hope this problem can be fixed early. Is there a workaround?

**Comment by onee-only** [comment]
> I discovered this problem when traversing multiple large git repos. I hope this problem can be fixed early. Is there a workaround?

Is MaxOpenDescriptors option set in your object storage?
I think setting it to some positive number should significantly decrease the memory allocation.

**Comment by MasonXon** [comment]
Not yet. I will test it in the next two days and report the results here.

**Comment by onee-only** [comment]
> After testing for a while, I changed max open files from 65535 to 4000, and the impact on memory does not seem to be very large.

Did you change the option value like the code snippet below?

```go
func main() {
	// ...
	
	storage := filesystem.NewStorageWithOptions(fs, cache.NewObjectLRUDefault(), filesystem.Options{
		MaxOpenDescriptors: <the value>,
	})

	// ...
}
```

**Comment by MasonXon** [comment]
No, I adjusted it directly through syscall, does it not work? But there is no storage related definition in my code, can you give a complete example so that I can adjust it?

**Comment by onee-only** [comment]
I didn't mean that you should set max open files of OS. I meant that you should set `MaxOpenDescriptors` when you create a storage like the code snippet above when you initialize/open a repository.

I don't know how you open a repository in the code. So I'll just show you possible options.
```go
package main

import (
	"github.com/go-git/go-billy/v5/osfs"
	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/plumbing/cache"
	"github.com/go-git/go-git/v5/storage/filesystem"
)

func main() {
	currentDir := osfs.New(".")

	dotGitDir, err := currentDir.Chroot(git.GitDirName)
	if err != nil {
		panic(err)
	}

	storage := filesystem.NewStorageWithOptions(dotGitDir, cache.NewObjectLRUDefault(), filesystem.Options{
		MaxOpenDescriptors: 1,
	})


	// If you used to use git.PlainOpen(".")
	repo, err := git.Open(storage, currentDir)

	// If you used to use git.PlainInit(".")
	repo, err := git.Init(storage, currentDir)

	// If you used to use git.PlainClone(".", ...)
	repo, err := git.Clone(storage, currentDir, &git.CloneOptions{
		// existing options you already have
	})

}
```

**Comment by MasonXon** [comment]
<img width="827" alt="image" src="https://github.com/go-git/go-git/assets/86297429/d77e6202-4336-43aa-ad2d-2fffd2b13c39">
There doesn't seem to be any significant improvement. Above is where I create the `*git.Repository`, which I call in ShowCommitInfo.

<img width="1238" alt="image" src="https://github.com/go-git/go-git/assets/86297429/db3d48ac-ee93-4071-8384-e09d93da3dfc">

**Comment by MasonXon** [comment]
[git2json alloc_space.html.zip](https://github.com/user-attachments/files/15912543/git2json.alloc_space.html.zip)

Because of this problem, I tried to use commitIterators and channel to read all commit histories, but I encountered [other problems](https://github.com/go-git/go-git/issues/1121). It seems that concurrent operations on repositories are not possible in go-git.

**Comment by onee-only** [comment]
<img width="1044" alt="image" src="https://github.com/go-git/go-git/assets/103808587/b63ff171-0620-4809-804d-07c7f945c8b1">


I think it is a progress that it actually does not allocate new reader repeatedly when we call 
`filesystem.(*ObjectStroage).EncodedObject` anymore. I can't determine how much it improved because those two allocation profiles seem like recorded on different situations.

And yes, go-git is not safe for concurrent operations. https://github.com/go-git/go-git/issues/773

**Comment by MasonXon** [comment]
Do you have any better suggestions or methods to solve the memory consumption problem of traversing all commit history in a large repository?

**Comment by onee-only** [comment]
I can't find any other workaround at this point. 
I think the only way to significantly decrease the memory allocation is to optimize DecodeTree.
I'm sorry that I can't help you more.

**Comment by MasonXon** [comment]
okay, thank a lot

**Comment by ironsweet** [comment]
I can reproduce the memory corruption case in commit patch. And setting MaxOpenDescriptors did fix the problem.

**Comment by ironsweet** [comment]
I can reproduce the memory corruption case. And setting MaxOpenDescriptors does not help.

**Comment by pjbgf** [comment]
Please note that the version in the branch `v6-exp` has some memory optimisations - packfile handling for example should consume less memory. Would you mind giving it a try with the latest from that branch and confirm whether that works for you?

For future reference, please try to share a snippet of the code used and a public repository where the issue could be reproduced - e.g. against `kubernetes/kubernetes`.
